Cleanup ext storage CRUD code, introduce storage config id

PVince81 opened the pull request:
Tried to make the external storage app code more meaningful.

But I hit against hard blocks like this: https://github.com/owncloud/core/blob/master/apps/files_external/js/settings.js#L136
Calling "removeMountPoint.php" for EVERY user ?

Also the lack of "id" makes it impossible to continue.
Will try and introduce a proper id in a separate PR first...

PVince81 commented on apps/files_external/appinfo/app.php:
```diff
@@ -6,26 +6,32 @@
-OCP\App::registerAdmin('files_external', 'settings');
-if (OCP\Config::getAppValue('files_external', 'allow_user_mounting', 'yes') == 'yes') {
-	OCP\App::registerPersonal('files_external', 'personal');
+\OC::$CLASSPATH['OC\Files\Storage\StreamWrapper'] = 'files_external/lib/streamwrapper.php';
```
Maybe another time, I need to focus on the essential :smile:

LukasReschke commented on apps/files_external/appinfo/application.php:
```diff
@@ -0,0 +1,56 @@
+
+
+}
+
```
You can do the automatic DI for that nowadays. – Either take a look at our docs (it's somewhere documented there) or ping me in IRC ;-)

LukasReschke replied:
=> We can even get rid of that file

PVince81 replied:
Ah yes, automatic DI. I completely forgot that (spent too much time away from the app framework lately...)

PVince81 commented:
@DeepDiver1975 some app framework pieces are available on master, in case you're still interested.

I'll continue this PR with the focus on mountpoint management (add/update/delete) based on the new storage_id setting that was introduced in 8.0. This would be to fix: https://github.com/owncloud/core/issues/14058 (prerequisite for storage options)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9398a9fb-ab7f-4a3f-ba00-e5bb552efb7c) was created.

PVince81 commented:
Just done a major cleanup of the JS side part.

#### Tasks
- [ ] write JS unit tests
- [ ] write service class to load/save mount points + unit test
- [ ] wire service class to OC_MountConfig for reading/saving the config to file (something to refactor in a future PR)
- [ ] fix/finish update spinner

#### Regression Testing
##### System mount
- [ ] Add new storage (storage must receive id)
- [ ] Update storage
- [ ] Delete storage
- [ ] Adding/removing users/groups
- [ ] Case "applicable for all"
- [ ] Renaming mount point
- [ ] Validation (when required fields are missing)
- [ ] Mountpoint conflicts
- [ ] Loading/deletion spinner

##### Personal mount
- [ ] Add new storage (storage must receive id)
- [ ] Update storage
- [ ] Delete storage
- [ ] Renaming mount point
- [ ] Validation (when required fields are missing)
- [ ] Mountpoint conflicts
- [ ] Loading/deletion spinner

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9398a9fb-ab7f-4a3f-ba00-e5bb552efb7c) completed: **4 new issues, 5 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b6d214bc-f728-4971-b24e-c031d5a9ab8c) was created.

PVince81 commented:
It's already looking good :smile: 
But there is still some work to be done.

PVince81 commented on apps/files_external/appinfo/routes.php:
```diff
@@ -22,28 +22,29 @@
+	$this,
+	array(
+		'resources' => array(
+			'storages' => array('url' => '/storages')
```
@LukasReschke @DeepDiver1975 @icewind1991 do you think we should use separate endpoints for "system-wide" mounts and "personal" mounts ?

LukasReschke replied:
I'd be in favour of that if it does not lead to too much duplication as we then can do the security checks via annotations and don't have to do those ugly "is user is admin then this otherwise that" stuff :)

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b6d214bc-f728-4971-b24e-c031d5a9ab8c) completed: **39 new issues, 18 updated code elements**

PVince81 commented:
BAD NEWS: due to the flexibility we offered before like putting value like "/somedir/$user/" in config fields, this would cause a single config entry to produce multiple entries in the "oc_storages" table.
This means that we **cannot** use the "storage_id" as a unique id for the mount point config :rage: 

It looks like we'll need to introduce yet another id, call it "external_storage_config_id" (or shorter) which would match every entry visible in the config in the admin pageWeb UI.
Obviously even with that approach the "mount.json" file will still have multiple entries, but those entries would all have "external_storage_config_id". In the future we can have that in the database.

PVince81 commented:
In principle this also means that we need a 1-N mapping table that maps "external_storage_config_id" to "oc_storages.numeric_id". Without this we cannot reliably delete the matching "oc_storages" entries on deletion in such situations where the "$user" field is used (#11544)

PVince81 commented:
and as long as we don't have "mount.json" in the database, we cannot reliably generate incremental "external_storage_config_id" because concurrency could make it generate twice the same.

Unless "external_storage_config_id" is a hash of the config... ARGHH

I'll leave this task aside for now until I find new ideas that does not involve moving mount.json to the database (which is a big task, especially regarding the migration + import scripts)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d94814d2-70fd-4bf3-8bfe-71434042c66a) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d94814d2-70fd-4bf3-8bfe-71434042c66a) completed: **19 new issues, 20 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ad93fdd2-e473-46a6-b234-616d243759a9) was created.

Xenopathic commented:
Well status checking is now mostly implemented. I'm still bringing over changes from #13515, like the async ability. It current lies a bit about the storage status, since when creating the storage there is no ID available and the storage status cannot be determined. But when IDs are implemented that will clear itself up.

Also, we now have the structure available for multiple error types. So after the troubles of this PR are sorted out (I'm in favour of moving `mount.json` to the database, maybe for 8.2 though), we can give the user better error information.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ad93fdd2-e473-46a6-b234-616d243759a9) completed: **13 new issues, 20 updated code elements**

PVince81 commented:
Thanks a lot @Xenopathic.

I think I wasn't done completely with the case "has no idea yet at creation time" because that's where I bumped into the issue on how to generate the storage_id, and that's when I realized that it wasn't unique/consistent... so for that part we need to sort out what to do.

For now I suggest to ignore concurrency and either use max(id) + 1 based on the existing config, or use a config based hash as a string id (like we kind of had before to detect how to group configs). This would be the temp solution until we can move to the DB.

Xenopathic commented:
@PVince81 only problem with using a hash of the config as the config ID is that the hash will change any time the config changes, so there is no point to the 'update' action.

Given that we don't want to move completely to the database just yet, my vote goes for the max(id) + 1 option.

PVince81 commented:
Ok, thanks for your input :smile: 

The problem with moving with the DB is that it will take a lot of coding, but also will require migration + import script (to import mount.json into the DB), which is quite a big task.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/92243c81-33e5-4004-b032-bf9c27a56f83) was created.

PVince81 commented:
I split the controller and service into two set of classes, one for global storages and one for user storages (aka personal mounts).
Also, now it generates an id.

Personal mount part is still broken at the moment, will look into that next.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/92243c81-33e5-4004-b032-bf9c27a56f83) completed: **4004 Issues, 123 Patches**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/92243c81-33e5-4004-b032-bf9c27a56f83) completed: **4004 Issues, 123 Patches**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1f50ffcd-0715-44d8-9508-2b0f8dc0975d) was created.

PVince81 commented:
@Xenopathic do you mind helping testing the "priority" field, you know it better than anyone how it works :smile: 
I tried to put it in to make sure it's kept, but I'm not sure it works. Also, did the "priority" field get lost in older versions when saving the configs ?

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1f50ffcd-0715-44d8-9508-2b0f8dc0975d) completed: **4004 Issues, 123 Patches**

Xenopathic commented:
@PVince81 Sure, I'll give it a test. Priority was never persistent if the storage config changed, it was just a way for the sysadmin to prioritise one storage over another with the same mount path when modifying mount.json directly.

PVince81 commented:
No, I don't think it should be a jQuery plugin because it's not a UI feature.
The fact that we use $tr is already bad enough.

But making OC.MountConfig a real non-static class would be good. But I have limited time to finish this feature so I might leave it as is for now.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1c8366e4-c7be-47d3-a93e-d06782df08fa) was created.

LukasReschke commented on apps/files_external/controller/globalstoragescontroller.php:
```diff
@@ -0,0 +1,143 @@
+use \OCP\IRequest;
+use \OCP\AppFramework\Http\DataResponse;
+use \OCP\AppFramework\Controller;
+use OC\AppFramework\Http;
```
Please use  the public pendant: `\OCP\AppFramework\Http`

LukasReschke commented on apps/files_external/controller/storagescontroller.php:
```diff
@@ -0,0 +1,136 @@
+use \OCP\IRequest;
+use \OCP\AppFramework\Http\DataResponse;
+use \OCP\AppFramework\Controller;
+use OC\AppFramework\Http;
```
Please use the public pendant: `\OCP\AppFramework\Http`

LukasReschke commented on apps/files_external/controller/userstoragescontroller.php:
```diff
@@ -0,0 +1,171 @@
+use \OCP\IRequest;
+use \OCP\AppFramework\Http\DataResponse;
+use \OCP\AppFramework\Controller;
+use OC\AppFramework\Http;
```
Please use the public pendant: `\OCP\AppFramework\Http`

PVince81 commented:
Added unit tests for the service classes.

I'll probably add some for the controllers too (and also handle the hooks case)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4d39feef-e07a-461f-a1b7-a14d68c0fdb0) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1c8366e4-c7be-47d3-a93e-d06782df08fa) completed: **4031 Issues, 124 Patches**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1c8366e4-c7be-47d3-a93e-d06782df08fa) completed: **4031 Issues, 124 Patches**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4d39feef-e07a-461f-a1b7-a14d68c0fdb0) completed: **4031 Issues, 124 Patches**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3b5f4393-b0dc-4a84-bf91-3fbf15afaa13) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3b5f4393-b0dc-4a84-bf91-3fbf15afaa13) completed: **4047 Issues, 124 Patches**

PVince81 commented:
- [ ] TODO: add thumbnail (and other options if available) to a new dropdown in the UI

PVince81 commented:
- [ ] TODO: set priority default to 100 (seems to be like that in the old code) instead of null / nothing

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d8e746d1-9774-494f-8558-915907416e2e) was created.

PVince81 commented:
- [ ] TODO: compare hook behavior with core (especially regarding config changes that do not affect $applicable or the mount point)

PVince81 commented:
Phew, I added hook stuff, look at this beauty of engineering: https://github.com/owncloud/core/commit/57826392765069435eedab665e1e7495850d6247#diff-24d8ac8577b2eb14422dbfbdd148fb61R145 (which comes with unit tests too).
Still not sure whether we should keep those hooks, they are currently only used to trigger etag propagation.
Each hook used to be fired for every change in $applicable. Renaming the mount point would trigger mount point deletion then recreation. I'm not sure we need exactly that for etag propagation, especially having such complex logic...

PVince81 commented:
Just checked master and saw that no hook is triggered for pure config changes (not rename and not $applicable). Same behavior on this branch.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d8e746d1-9774-494f-8558-915907416e2e) completed: **4054 Issues, 125 Patches**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d2a043a9-2996-454a-8017-3aa5e237f260) was created.

PVince81 commented:
I decided to replace the magic "array" with a real class called "StorageConfig" which contains the external storage config. It helps enforce a specific format. Because I'm a bit tired of those too flexible arrays...

icewind1991 commented on apps/files_external/js/settings.js:
```diff
@@ -31,153 +35,371 @@ function highlightInput($input) {
+	 *
+	 * @type Object.<string,object>
+	 */
+	backendOptions: {},
```
This doesnt work as expected, since a new instance of a class is a shallow copy of the prototype each instance will share the same `backendOptions`, instead you need to set `this.backendOptions={}` in the constructor

PVince81 replied:
True... got confused with PHP again. I was told to do it this way in PHP while I used the JS argument. Thanks for pointing out.

icewind1991 commented on apps/files_external/js/settings.js:
```diff
@@ -31,153 +35,371 @@ function highlightInput($input) {
+	 *
+	 * @type Array.<string>
+	 */
+	applicableUsers: [],
```
same as above

icewind1991 commented on apps/files_external/service/storagesservice.php:
```diff
@@ -0,0 +1,309 @@
+	 * @param StorageConfig $storage storage data
+	 * @param string $signal signal to trigger
+	 */
+	protected function triggerHooks(StorageConfig $storage, $signal) {
```
```php
abstract protected function triggerHooks(StorageConfig $storage, $signal);
```

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d2a043a9-2996-454a-8017-3aa5e237f260) completed: **4054 Issues, 125 Patches**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/88c58f42-bc16-45bd-bf9a-4e2f65fa45c8) was created.

PVince81 commented:
Fixed @icewind1991's comments.
I also moved around/refactored the JS classes a bit in preparation for writing JS unit tests (like putting global event handlers into the MountConfigListView class)

icewind1991 commented on apps/files_external/controller/storagescontroller.php:
```diff
@@ -0,0 +1,157 @@
+use \OCA\Files_external\NotFoundException;
+use \OCA\Files_external\Lib\StorageConfig;
+
+abstract class StoragesController extends Controller {
```
shouldn't this have an abstract `create` and `update` method?

PVince81 replied:
Not really because they have a different signature that matches the app framework injection.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/88c58f42-bc16-45bd-bf9a-4e2f65fa45c8) completed: **No issues found**

PVince81 commented:
- [ ] BUG: returned backend options are wrong (but has no effect in the UI as it's not used)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/34d40133-5f14-4b6e-b6b2-46b4b394eb74) was created.

PVince81 commented:
Behold, the first JS unit tests ever written for files_external GUI: https://github.com/owncloud/core/commit/552f39a57e6bff089ab972f429cc6203e7885b55 ! :smile:

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/34d40133-5f14-4b6e-b6b2-46b4b394eb74) completed: **4060 Issues, 126 Patches**

PVince81 commented:
- [ ] test migrating from old configs / need to generate missing id

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/30aa4758-8d1c-45b3-8c2f-3e1246c8e3ea) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/30aa4758-8d1c-45b3-8c2f-3e1246c8e3ea) completed: **4062 Issues, 126 Patches**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/dad0adef-d740-43a2-bfc6-e770f5290e29) was created.

Xenopathic commented:
Right, priorities are now seen client-side, in a hidden input. This means that the infrastructure is there for the ability to change a storage priority from the GUI (which is definitely an 'advanced' feature, and should be hidden in a drop down)

PVince81 commented:
Thanks a lot @Xenopathic

PVince81 commented:
Trying to run/fix the unit tests when files_external is enabled, but am getting weird shit (even on master!):
```
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 6720 bytes) in /srv/www/htdocs/owncloud/lib/private/image.php on line 486
PHP Stack trace:
PHP   1. {main}() /home/vincent/opt/bin/phpunit:0
PHP   2. PHPUnit_TextUI_Command::main() /home/vincent/opt/bin/phpunit:620
PHP   3. PHPUnit_TextUI_Command->run() phar:///home/vincent/opt/bin/phpunit/phpunit/TextUI/Command.php:103
PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///home/vincent/opt/bin/phpunit/phpunit/TextUI/Command.php:151
PHP   5. PHPUnit_Framework_TestSuite->run() phar:///home/vincent/opt/bin/phpunit/phpunit/TextUI/TestRunner.php:388
PHP   6. PHPUnit_Framework_TestSuite->run() phar:///home/vincent/opt/bin/phpunit/phpunit/Framework/TestSuite.php:716
PHP   7. PHPUnit_Framework_TestCase->run() phar:///home/vincent/opt/bin/phpunit/phpunit/Framework/TestSuite.php:716
PHP   8. PHPUnit_Framework_TestResult->run() phar:///home/vincent/opt/bin/phpunit/phpunit/Framework/TestCase.php:687
PHP   9. PHPUnit_Framework_TestCase->runBare() phar:///home/vincent/opt/bin/phpunit/phpunit/Framework/TestResult.php:609
PHP  10. PHPUnit_Framework_TestCase->runTest() phar:///home/vincent/opt/bin/phpunit/phpunit/Framework/TestCase.php:731
PHP  11. ReflectionMethod->invokeArgs() phar:///home/vincent/opt/bin/phpunit/phpunit/Framework/TestCase.php:846
PHP  12. OC\Core\Avatar\AvatarControllerTest->testGetAvatar() phar:///home/vincent/opt/bin/phpunit/phpunit/Framework/TestCase.php:846
PHP  13. OC_Image->__construct() /srv/www/htdocs/owncloud/tests/core/avatar/avatarcontrollertest.php:140
PHP  14. OC_Image->load() /srv/www/htdocs/owncloud/lib/private/image.php:71
PHP  15. OC_Image->loadFromFile() /srv/www/htdocs/owncloud/lib/private/image.php:437
PHP  16. imagecreatefromjpeg() /srv/www/htdocs/owncloud/lib/private/image.php:486
```
It randomly fails, unless I delete the whole folder "apps/files_external/tests/backends".

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/dad0adef-d740-43a2-bfc6-e770f5290e29) completed: **4062 Issues, 126 Patches**

PVince81 commented:
The unit test memory issue mysteriously disappeared...

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/af50a5ac-fe37-4d8f-b0a5-57160737f6d5) was created.

PVince81 commented:
Nope, it didn't, I forgot to enable "files_external" while running the tests...

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/af50a5ac-fe37-4d8f-b0a5-57160737f6d5) completed: **4062 Issues, 126 Patches**

PVince81 commented:
Seems someone fixed it parly on master, going to rebase this branch onto master.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/502c2153-55b0-4bcd-af97-fe6996299b55) was created.

PVince81 commented:
@Xenopathic looks like your "status" change breaks the mountconfig.php tests. Now the ugly old legacy `addMountConfig` returns `0` instead of `true`, because it uses the status.

I'll see if I can get rid of "addMountConfig" completely.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/502c2153-55b0-4bcd-af97-fe6996299b55) completed: **65 new issues, 124 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/724ccec3-9d2b-4baa-9578-2350f0521baa) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/724ccec3-9d2b-4baa-9578-2350f0521baa) completed: **65 new issues, 127 updated code elements**

Xenopathic commented:
Suggestion: we turn `priority` into a mount option?

PVince81 commented:
Yes, but it means the format will need to change because now it's not on the same level. Which means we need a repair step to convert to config. Or read it from both places.

PVince81 commented:
- [ ] Potential bug: currently when the new Service class reads the legacy config, it would group everything by mount point. But the configs could be different. Might need to reuse the grouping logic from https://github.com/owncloud/core/blob/master/apps/files_external/lib/config.php#L360 instead, and give each group a different id

PVince81 commented:
- [ ] add unit test to read different forms of legacy configs (with/without redundancy, etc)

LukasReschke commented on apps/files_external/controller/globalstoragescontroller.php:
```diff
@@ -0,0 +1,151 @@
+	 * @param string $appName
+	 * @param IRequest $request
+	 * @param IConfig $config
+	 * @param GlobalStoragesService $globalStoragesService
```
PHPDoc is invalid

LukasReschke replied:
1. Wrong casing for appname
2. Misses IL10N

LukasReschke commented on apps/files_external/controller/storagescontroller.php:
```diff
@@ -0,0 +1,157 @@
+	 * @param string $appName
+	 * @param IRequest $request
+	 * @param IConfig $config
+	 * @param StoragesService $storagesService
```
Also invalid PHPDoc

LukasReschke commented on apps/files_external/controller/userstoragescontroller.php:
```diff
@@ -0,0 +1,178 @@
+	 * @param string $appName
+	 * @param IRequest $request
+	 * @param IConfig $config
+	 * @param UserStoragesService $userStoragesService
```
Also invalid PHPDoc

PVince81 replied:
Invalid in what way ? Incomplete or do I need to use the full namespace ?

LukasReschke replied:
Same as https://github.com/owncloud/core/pull/11884#discussion_r26391833

IDE helps:
![2015-03-13_16-01-08](https://cloud.githubusercontent.com/assets/878997/6640821/2b7a80a8-c99a-11e4-95b6-5461c09b9a56.png)

LukasReschke commented on apps/files_external/js/settings.js:
```diff
@@ -31,298 +35,511 @@ function highlightInput($input) {
+			return element.name;
+		},
+		formatResult: function (element) {
+			var $result = $('<span><div class="avatardiv"/><span>'+escapeHTML(element.displayname)+'</span></span>');
```
I'd prefer if we do something like `$('span .avatardiv').text('foo')`

Such string concatenation really has to die :see_no_evil:

PVince81 replied:
not touching this code here. I only moved this function up in the file.

LukasReschke commented on apps/files_external/js/settings.js:
```diff
@@ -31,298 +35,511 @@ function highlightInput($input) {
+				.attr('data-displayname', element.displayname);
+			if (element.type === 'group') {
+				var url = OC.imagePath('core','places/contacts-dark'); // TODO better group icon
+				$div.html('<img width="32" height="32" src="'+url+'">');
```
Can we use JQuery instead of string concatenation here as well?

PVince81 replied:
or handlebars :-P

LukasReschke commented on apps/files_external/js/settings.js:
```diff
@@ -31,298 +35,511 @@ function highlightInput($input) {
+		},
+		formatSelection: function (element) {
+			if (element.type === 'group') {
+				return '<span title="'+escapeHTML(element.name)+'" class="group">'+escapeHTML(element.displayname+' '+t('files_external', '(group)'))+'</span>';
```
Same for those

LukasReschke commented on apps/files_external/js/settings.js:
```diff
@@ -31,298 +35,511 @@ function highlightInput($input) {
-						}
-					}
-				});
+				return '<span title="'+escapeHTML(element.name)+'" class="user">'+escapeHTML(element.displayname)+'</span>';
```
And those

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/6f713b64-2e33-4a1a-81a2-6f1da7b439c4) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/6f713b64-2e33-4a1a-81a2-6f1da7b439c4) completed: **68 new issues, 128 updated code elements**

Xenopathic commented:
I thought about moving `priority` into `mountOptions` some more, and have decided against it.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ce86f7eb-617f-44cf-83cd-434b2170523b) was created.

PVince81 commented:
Phew. Now when reading legacy configs, it will generate config ids. For this it will use the config hash logic (which we had before) to find out which configs belong together (ex: applicable for multiple groups/users)

The config hash is now also based on the priority and mount options (in case some funny person decided to have two mount points with the same name but different users and different mount options)

PVince81 commented:
Okay, the groundwork needed to put the mount options in the UI is done.

Next step: add the new UI. (unless you guys want to merge this first)

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ce86f7eb-617f-44cf-83cd-434b2170523b) completed: **No issues found**

PVince81 commented:
I'm marking this PR as "To review" because adding the mount options dropdown will bring its own issues and delays (like design reviews). So I'll make a separate PR that is based on this one.

Please review.

Jenkins PR will follow.

LukasReschke commented on apps/files_external/controller/storagescontroller.php:
```diff
@@ -0,0 +1,157 @@
+	protected $service;
+
+	/**
+	 * @param string $appName
```
Still invalid :see_no_evil:

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d995cfad-d215-4060-8315-08ccf17761ec) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d995cfad-d215-4060-8315-08ccf17761ec) completed: **61 new issues, 132 updated code elements**

PVince81 commented:
Finally green light !

Please review @DeepDiver1975 @icewind1991 @LukasReschke @nickvergessen

PVince81 commented:
Test plan here: https://github.com/owncloud/core/pull/11884#issuecomment-73747464

LukasReschke commented on apps/files_external/js/settings.js:
```diff
@@ -31,298 +35,511 @@ function highlightInput($input) {
+/**
+ * Initialize select2 plugin on the given elements
+ *
+ * @param {Array<Object>} array of jQuery elements
```
JSDoc noob  here. Don't we need to state `$elements` here as well?

To my defense: I tried `buildjsdocs.sh` but it bails out with:

```
➜  master git:(ext-appframework) bash buildjsdocs.sh
unbuild jsdoc@3.3.0-beta2
/Users/lukasreschke/Documents/Programming/master/build/node_modules/jsdoc -> /Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc
jsdoc@3.3.0-beta2 build/node_modules/jsdoc
/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/tag.js:33
    return meta && meta.code && meta.code.name === text && text.match(/(?:^\s+
                                                                ^
TypeError: undefined is not a function
    at mustPreserveWhitespace (/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/tag.js:33:65)
    at trim (/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/tag.js:43:10)
    at new exports.Tag (/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/tag.js:189:17)
    at Doclet.addTag (/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/doclet.js:199:18)
    at newSymbolDoclet (/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/src/handlers.js:256:19)
    at null.<anonymous> (/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/src/handlers.js:318:13)
    at emit (events.js:110:17)
    at Visitor.visitNode (/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/src/visitor.js:281:16)
    at Visitor.visit (/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/src/visitor.js:191:27)
    at Walker.recurse (/Users/lukasreschke/Documents/Programming/master/build/lib/node_modules/jsdoc/lib/jsdoc/src/walker.js:531:44)
```

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	/**
+	 * Sets the configuration id
+	 *
+	 * @param int configuration id
```
`@param int $id configuration id`

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	 * "files" folder.
+	 * The path will be normalized.
+	 *
+	 * @param string path
```
![2015-03-16_18-37-38](https://cloud.githubusercontent.com/assets/878997/6672264/92bdd832-cc0b-11e4-8ec1-280c27a36365.png)

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	/**
+	 * Returns the external storage backend class name
+	 *
+	 * @return string external storage backend class name
```
![2015-03-16_18-38-16](https://cloud.githubusercontent.com/assets/878997/6672273/a49b280c-cc0b-11e4-8fb2-329a0b0b275d.png)

DeepDiver1975 replied:
@PVince81 wana give PHPStorm a try? Really helps as you can see! :speak_no_evil:

PVince81 replied:
Well I can see this here too, just need to decondition myself to not ignore warnings any more.
![vim-phpcs](https://cloud.githubusercontent.com/assets/277525/6684845/3d0c02a8-cc95-11e4-9da5-0506c157ccfb.png)

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	/**
+	 * Sets the external storage backend class name
+	 *
+	 * @param string external storage backend class name
```
As well

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	/**
+	 * Sets the mount priotity
+	 *
+	 * @param int priority
```
as well

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	/**
+	 * Sets the users for which to mount this storage
+	 *
+	 * @param array applicable users
```
as well

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	 *
+	 * @param array applicable users
+	 */
+	public function setApplicableUsers($applicableUsers) {
```
Type-hint `array $applicableUsers`?

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	/**
+	 * Sets the groups for which to mount this storage
+	 *
+	 * @param array applicable groups
```
as well

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	 *
+	 * @param array applicable groups
+	 */
+	public function setApplicableGroups($applicableGroups) {
```
Type-hint `array $applicableGroups` ?

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	/**
+	 * Sets the mount-specific options
+	 *
+	 * @param array applicable groups
```
as well

LukasReschke replied:
also pretty sure that's not the groups?

LukasReschke commented on apps/files_external/lib/storageconfig.php:
```diff
@@ -0,0 +1,272 @@
+	 *
+	 * @param array applicable groups
+	 */
+	public function setMountOptions($mountOptions) {
```
`array $mountOptions`?

LukasReschke commented on apps/files_external/appinfo/routes.php:
```diff
@@ -22,28 +22,30 @@
 namespace OCA\Files_External\Appinfo;
 
+/**
+ * @var $this OC\Route\Router
```
` \OC\Route\Router` - will cause problems with the namespace otherwise

LukasReschke commented on apps/files_external/controller/userstoragescontroller.php:
```diff
@@ -0,0 +1,181 @@
+	 *
+	 * @return DataResponse
+	 */
+	public function update(
```
Is this missing a `@NoAdminRequired`?

PVince81 replied:
Yes, added it

LukasReschke commented:
What about the `index` methods as well for those?

Now accessing `/index.php/apps/files_external/globalstorages` will simply yield a fatal PHP error

PVince81 commented:
I can add an index route but it won't be used by anything yet, until we port the UI too (hopefully in a future PR)

LukasReschke commented:
> I can add an index route but it won't be used by anything yet, until we port the UI too (hopefully in a future PR)

Then let's not add them for now :)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d0f0e566-df9f-4177-a754-e6fb2cca8794) was created.

PVince81 commented:
Fixed the PHPdoc and added the missing NoAdminRequired tag

DeepDiver1975 commented:
@owncloud-bot retest this please

icewind1991 commented:
Looks good, basic testing works fine :+1:

PVince81 commented:
Needs a second reviewer @Xenopathic @DeepDiver1975 @LukasReschke 
(I agree with the changes from @Xenopathic in this PR)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/af71db3d-9252-4a43-a23f-ad1482cf6781) was created.

Xenopathic commented:
Eek, I forgot to pull across async status checking from #13515. That last commit implements that - very simple, the framework was already there. Please review @icewind1991 ? Thanks

PVince81 commented:
Tested @Xenopathic's async check, works. This :+1: counts for the commit https://github.com/owncloud/core/commit/c37913b1d506c0a4afe035de0e4c4fd5a80bb6ea

Will raise a Jenkins PR...

Xenopathic commented:
Likewise, I've tested (and developed on) the rest of this PR done by @PVince81, so :+1: from me. I'd suggest one more independent reviewer though. @icewind1991 already reviewed all but that last commit, so all we really need is one final review...

Xenopathic commented:
It'd be great to get #13641 merged at the same time, so that we get full support for downed storages.

icewind1991 commented:
:+1: for the async check

PVince81 commented:
Thanks a lot guys.

Xenopathic commented:
Now starts the process of rebasing every other PR that so much as looks at files_external code :laughing:

MorrisJobke commented:
@PVince81 https://ci.owncloud.org/job/server-master-linux-externals-smb-windows/database=sqlite,external=smb-windows,label=master/205/console <- caused by this PR or just sun winds? cc @DeepDiver1975
